File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    answer = -1

    if target not in words:
        return 0

    visited=[0]*(len(words))
    
    q = deque()
    q.append(begin)

    while q:
        w = q.popleft()
        k = words.index(w) if begin!=w else 0
        for i in range(len(words)):
            if visited[i] != 0:
                continue
            cnt = 0
            for j in range(len(w)):
                if words[i][j] != w[j]:
                    cnt+=1
            if cnt == 1: # 한글자만 다르다면
                q.append(words[i])
                visited[i] = visited[k]+1        

    return visited[words.index(target)]

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;
class Solution {
    public static int solution(String begin, String target, String[] words) {

        if(!Arrays.asList(words).contains(target)){
            return 0;
        }

        int[] dist = new int[words.length];

        for(int k=0; k<words.length; k++){
            dist[k] = 0;
        }

        Deque<String> deque = new ArrayDeque<>();
        deque.add(begin);


                
        while(!deque.isEmpty()){
            String word = deque.pop();
            int idx = Arrays.asList(words).indexOf(word);

            if(word.equals(target)){
                return dist[idx];
            }

            for(String str : words){
                int cnt = 0;
                int sIdx = Arrays.asList(words).indexOf(str);

                for(int i=0; i<str.length(); i++){
                    if(word.charAt(i) != str.charAt(i)){
                        cnt += 1;
                    }
                }
                if (cnt == 1 && dist[sIdx]==0){
                    deque.add(str);
                    dist[sIdx] = idx==-1? 1 : dist[idx]+1;
                }
            }
        }
        return 0;
    }
}